pkg/providers/docker/controlplane_test.go

```diff
@@ -0,0 +1,527 @@
+	g := NewWithT(t)
+	logger := test.NewNullLogger()
+	ctx := context.Background()
+	spec := test.NewFullClusterSpec(t, testClusterConfigMainFilename)
+	client := test.NewFakeKubeClientAlwaysError()
+	_, err := docker.ControlPlaneSpec(ctx, logger, client, spec)
+	g.Expect(err).To(MatchError(ContainSubstring("updating docker immutable object names")))
+}
+
+func capiCluster() *clusterv1.Cluster {
+	return &clusterv1.Cluster{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Cluster",
+			APIVersion: "cluster.x-k8s.io/v1beta1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test",
+			Namespace: "eksa-system",
+		},
+		Spec: clusterv1.ClusterSpec{
+			ClusterNetwork: &clusterv1.ClusterNetwork{
+				APIServerPort: nil,
+				ServiceDomain: "cluster.local",
+				Services: &clusterv1.NetworkRanges{
+					CIDRBlocks: []string{"10.96.0.0/12"},
+				},
+				Pods: &clusterv1.NetworkRanges{
+					CIDRBlocks: []string{"192.168.0.0/16"},
+				},
+			},
+			ControlPlaneRef: &corev1.ObjectReference{
+				Kind:       "KubeadmControlPlane",
+				Name:       "test",
+				Namespace:  "eksa-system",
+				APIVersion: "controlplane.cluster.x-k8s.io/v1beta1",
+			},
+			ManagedExternalEtcdRef: &corev1.ObjectReference{
+				Kind:       "EtcdadmCluster",
+				Name:       "test-etcd",
+				Namespace:  "eksa-system",
+				APIVersion: "etcdcluster.cluster.x-k8s.io/v1beta1",
+			},
+			InfrastructureRef: &corev1.ObjectReference{
+				Kind:       "DockerCluster",
+				Name:       "test",
+				Namespace:  "eksa-system",
+				APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+			},
+		},
+	}
+}
+
+func dockerCluster() *dockerv1.DockerCluster {
+	return &dockerv1.DockerCluster{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "DockerCluster",
+			APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test",
+			Namespace: constants.EksaSystemNamespace,
+		},
+	}
+}
+
+func dockerMachineTemplate(name string) *dockerv1.DockerMachineTemplate {
+	return &dockerv1.DockerMachineTemplate{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "DockerMachineTemplate",
+			APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: constants.EksaSystemNamespace,
+		},
+		Spec: dockerv1.DockerMachineTemplateSpec{
+			Template: dockerv1.DockerMachineTemplateResource{
+				Spec: dockerv1.DockerMachineSpec{
+					CustomImage: "public.ecr.aws/l0g8r8j6/kubernetes-sigs/kind/node:v1.19.8-eks-d-1-19-4-eks-a-0.0.1.build.38",
+					ExtraMounts: []dockerv1.Mount{
+						{
+							ContainerPath: "/var/run/docker.sock",
+							HostPath:      "/var/run/docker.sock",
+							Readonly:      false,
+						},
+					},
+					Bootstrapped: false,
+				},
+			},
+		},
+	}
+}
+
+func kubeadmControlPlane() *controlplanev1.KubeadmControlPlane {
+	var kcp *controlplanev1.KubeadmControlPlane
+	b := []byte(`apiVersion: controlplane.cluster.x-k8s.io/v1beta1
```

Comment:
can we add this as an expected results cp yaml in the testdata folder? in my opinion its easier to maintain the yaml and use for other tests.

Reply:
If we take a route like this, I would suggest maybe trying to leverage one of the existing expected results test files?

But I would vote to define the objects directly in the file as it is now. Imo, it's better for debugging & verifying the contents of each individual cp object in the test

Reply:
Is there's an existing way to parse a yaml into an object? It may be possible to cover both options.

I do think the yaml file is easier to do deal with in general, but it may be more beneficial for debugging and verification as is now like @taneyland mentioned.

Reply:
Also, since I used Kube123 for the tests, I defined and used a more up to date VersionsBundle object than the existing one in `docker_test.go` , so the result would never match one the existing files.

Not sure if the VersionsBundle matters as much when using a different Kube version for the tests though. Let me know what you guys think.

Reply:
Yeah good point, the versions bundle should match so using the existing files wouldn't work since we updated the version. I think we should keep it as is.

Reply:
Personally i think this makes the test files look more cluttered, and would be inconsistent compared to all the other tests we have written. I am fine with it however if the team decides to use objects from here on out.

Reply:
I probably missed part of the convo, but any reason not to use go structs directly?